lrlex/src/lib/parser.rs

```diff
@@ -287,9 +287,21 @@ impl<StorageT: TryFrom<usize>> LexParser<StorageT> {
             .find(|c| c == '\n')
             .unwrap_or(self.src.len() - i);
         let line = self.src[i..i + line_len].trim_end();
-        let rspace = match line.rfind(' ') {
-            Some(j) => j,
-            None => return Err(self.mk_error(LexErrorKind::MissingSpace, i)),
+        let rspace = match line.rfind(|ch| ch == '\'' || ch == '"') {
+            Some(x) => match line[0..x].rfind(|ch| ch == '\'' || ch == '"') {
```

Comment:
will this not match non-matching quoting?
e.g. it reads like
.          "a'test 'b"
will produce a _name_ of **b**

Reply:
I am starting to get the feeling that this is complex enough that it may be cleaner to switch to regex for rule parsing

Reply:
These mismatched quotes get caught in a later pass when the error `InvalidName` is produced, so while this will make it through this early pass looking for space, it'll fail later.

That is why i hadn't bothered to check that they actually matched, but I think you're probably right e.g. given that "'" appears to currently be a valid rule we should probably check for matching quotes here too.

Reply:
Other option: could we use the string parsing from the rust compiler, then we'd support **all** permissible rust strings, which may align with users' expectations.

Reply:
> could we use the string parsing from the rust compiler

Is that something that rustc exposes? [I'm not sure if it's a good idea or not, but my working guess is that either way, we can't actually get at it.]

Reply:
I'll have to think about it a bit, the difficulty in using both something like regex, or a string parsing crate is that this currently needs to be matched in reverse, because the regex itself can contain an unbounded number of ` ` characters.
such as in `([a-c] |[w-z] ) "foo"`, 

I don't see any regex API for matching in reverse, and it seems unlikely that any string parsing crate would have one.
Further is that regex also can contain quotes, so I'm a bit hesitant to do something to the effect of `^.*[ ]("[^"]*"|'[^']*')$`.
because it seems like it could match in the middle of a regex perhaps like `[ "] "space_or_quote"`.

So I think that this, being easiest while matching in reverse is probably the biggest hurdle to doing something more principled here.

Reply:
> So I think that this, being easiest while matching in reverse is probably the biggest hurdle to doing something more principled here.

That sounds plausible.

> I think you're probably right e.g. given that "'" appears to currently be a valid rule we should probably check for matching quotes here too.

Good idea.

Reply:
So, I believe this is resolved in https://github.com/softdevteam/grmtools/pull/335/commits/6606b79e505e9cc1f1d19eabbf44d54ada9b9156 there is some info in the commit message.

But essentially this attempt to match the space before the quotes turned out to just not be possible.
At the same time the additional error checking needed to cover `a "` can also cover `a " "`.

So I reverted back to just finding the right-most space, and added tests for some cases that the first patch in this series had broken by skipping over the rightmost space and finding one in a regex.

Reply:
I suppose one other thing which we could consider is requiring escaping for `"`, `'` and `' '` in a regex.
So these would become '\"`, etc. with that we could search forwards to find the first non-escaped space.

It may produce better errors for these invalid name situations, but less readable regular expressions and compatibility.
So there is definitely a trade off in doing so.

Reply:
This looks like a meaningful improvement to me -- @SMartinScottLogic ?